Userland/Libraries/LibC/complex.h

```diff
@@ -74,4 +75,64 @@ static inline long double(cimagl)(long double complex z)
     return cimagl(z);
 }
 
+// https://pubs.opengroup.org/onlinepubs/9699919799/functions/cabs.html
+static inline long double cabsl(long double complex z)
+{
+    // Algorithm 312: Absolute value and square root of a complex number
+    // https://dl.acm.org/doi/10.1145/363717.363780
+    long double real = creal(z);
+    long double imag = cimagl(z);
+
+    if (real == 0)
+        return imag;
+
+    if (imag == 0)
+        return real;
+
+    if (real > imag)
+        return real * sqrtl(1 + (imag / real) * (imag / real));
+
+    return imag * sqrtl(1 + (real / imag) * (real / imag));
+}
+
+static inline double cabs(double complex z)
+{
+    return cabsl(z);
+}
+
+static inline float cabsf(float complex z)
+{
+    return cabsl(z);
+}
+
+// https://pubs.opengroup.org/onlinepubs/9699919799/functions/csqrt.html
+static inline long double complex csqrtl(long double complex z)
+{
+    // Algorithm 312: Absolute value and square root of a complex number
+    // https://dl.acm.org/doi/10.1145/363717.363780
+    long double real = creal(z);
+    long double imag = cimagl(z);
+
+    if (real == 0 && imag == 0)
+        return CMPLXL(0, 0);
+
+    long double a = sqrtl((fabsl(real) + cabsl(z)) * 0.5);
+
+    if (real >= 0)
+        return CMPLXL(a, imag / (a + a));
+
+    long double b = (imag < 0) ? -a : a;
+    return CMPLXL(imag / (b + b), b);
```

Comment:
This looks like a faithful implementation of the paper, yes.

However, real and imag can still "round to zero" while computing `a`, for example when called with ridiculously small values like the subnormals. How about we don't check for `real == 0.0 && imag == 0.0`. Instead, I suggest:

```c++
    long double a_squared = (fabsl(real) + cabsl(z)) * 0.5;
    if (a_squared <= 0.0) { return CMPLXL(0, 0); }
    // Square-rooting a positive number makes it get closer to 1, so `a` is clearly non-zero.
    long double a = sqrtl(a_squared);
    // Continue as before
```

Can you also add tests to make sure that there were no little typos or other mistakes? Something like:
* 1
* -1
* 2
* 3i
* -9i
* -7+24i which should result in "exactly" 3+4i
* 7+24i which should result in "exactly" 4+3i
* -9+0.00001i which should be somewhat close to 0+3i
* -9-0.00001i which should be somewhat close to 0-3i. Note that if the implementation splits the sqrt-branch in a different place, this should be documented
* 1e-4950 `== std::numeric_limits<long double>::denorm_min()` in STL, I forgot how we do it in AK. The result should be something like `6.03755e-2476`.

I got the last two numbers using
```c++
// g++ foo.cpp -Wall -pedantic -o foo && ./foo
#include <cmath>
#include <iostream>
#include <limits>
int main() {
    std::cout << std::numeric_limits<long double>::denorm_min() << '\n';
    std::cout << sqrtl(std::numeric_limits<long double>::denorm_min()) << '\n';
}
```

Reply:
changed the logic, and added the tests